From d6a01d9055891be43faccca45eade93884f6d45b Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 2 Jan 2024 11:35:59 +0100
Subject: [PATCH 0290/1695] drivers: stm32_iac: add
 CFG_STM32_PANIC_ON_IAC_EVENT switch

Platform will panic if CFG_STM32_PANIC_ON_IAC_EVENT is set upon receiving
an IAC event. This switch is default enabled on STM32MP2x platforms.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 068714277e142ae42bcee1534b09292fc7cd3f19 drivers: stm32_iac: add CFG_STM32_PANIC_ON_IAC_EVENT switch]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Ife09af16a0504ae04e361630ed21a6ba547dd7c8
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/337892
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/arch/arm/plat-stm32mp2/conf.mk | 3 +++
 core/drivers/firewall/stm32_iac.c   | 7 +++++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/conf.mk b/core/arch/arm/plat-stm32mp2/conf.mk
index ad295355a..9e4da1221 100644
--- a/core/arch/arm/plat-stm32mp2/conf.mk
+++ b/core/arch/arm/plat-stm32mp2/conf.mk
@@ -90,3 +90,6 @@ ifeq ($(CFG_STM32MP25_RSTCTRL),y)
 $(call force,CFG_DRIVERS_RSTCTRL,y)
 $(call force,CFG_STM32_RSTCTRL,y)
 endif
+
+# Optional behavior upon receiving illegal access events
+CFG_STM32_PANIC_ON_IAC_EVENT ?= y
diff --git a/core/drivers/firewall/stm32_iac.c b/core/drivers/firewall/stm32_iac.c
index 2d63ab6c4..45c43e097 100644
--- a/core/drivers/firewall/stm32_iac.c
+++ b/core/drivers/firewall/stm32_iac.c
@@ -174,7 +174,7 @@ static enum itr_return stm32_iac_itr(struct itr_handler *h __unused)
 			isr &= io_read32(base + _IAC_IER0 + offset);
 		}
 
-		/* A legit IAC occurred, platform will panic */
+		/* A legit IAC occurred, platform may panic */
 		if (isr) {
 			do_panic = true;
 			EMSG("IAC exceptions [%d:%d]: %#x",
@@ -194,8 +194,11 @@ static enum itr_return stm32_iac_itr(struct itr_handler *h __unused)
 		}
 	}
 
-	if (do_panic)
+	if (do_panic) {
 		stm32_rif_access_violation_action();
+		if (IS_ENABLED(CFG_STM32_PANIC_ON_IAC_EVENT))
+			panic();
+	}
 
 	return ITRR_HANDLED;
 }
-- 
2.47.2

